pkg/operators/common/kubeinventorycache.go

```diff
@@ -110,10 +127,120 @@ func (cache *K8sInventoryCache) Stop() {
 	cache.useCount--
 }
 
+func (cache *K8sInventoryCache) removeOldPods() {
+	now := time.Now()
+	cache.oldPods.Range(func(key, value any) bool {
+		if now.Sub(value.(oldResource[*v1.Pod]).deletionTimestamp) > oldObjectTTL {
+			cache.oldPods.Delete(key)
+		}
+		return true
+	})
+}
+
+func (cache *K8sInventoryCache) removeOldSvcs() {
+	now := time.Now()
+	cache.oldSvcs.Range(func(key, value any) bool {
+		if now.Sub(value.(oldResource[*v1.Service]).deletionTimestamp) > oldObjectTTL {
+			cache.oldSvcs.Delete(key)
+		}
+		return true
+	})
+}
+
 func (cache *K8sInventoryCache) GetPods() ([]*v1.Pod, error) {
-	return cache.pods.List(labels.Everything())
+	cache.removeOldPods()
+
+	pods := make([]*v1.Pod, 0)
+	addedPods := make(map[string]any)
+	cache.pods.Range(func(_, value any) bool {
+		pods = append(pods, value.(*v1.Pod))
+		addedPods[getObjectKey(value)] = nil
+		return true
+	})
+	cache.oldPods.Range(func(_, value any) bool {
+		pod := value.(oldResource[*v1.Pod]).obj
+		if _, ok := addedPods[getObjectKey(pod)]; !ok {
+			pods = append(pods, pod)
+			// Not needed to add to addedPods, since there are no duplicate keys in a map
+		}
+		return true
+	})
```

Comment:
Hmm, the problem is at least possible.
When implementing I wanted to avoid any additional mutexes, but maybe we need one. I think there is no other way to keep two lists "in sync" :/

Reply:
Without mutex:
Instead of removing a pod from `oldPods` when the same namespace/podsname is used we can leave it in there and just add the newly added pod to `pods`.

So in the situation you described the previous old pod gets returned to Thread A. I think this behavior is acceptable

Reply:
TBH, I don't know the performance hit, if there is any. It might be even a little bit faster is there is almost no contention.

Reply:
There are probably ways but the problem with benchmarks is always: How realistic are they?

How many container create and deletion events would happen concurrently?
Meanwhile we have concurrent enrichments - of how many events per seconds?


But in the end I would say a mutex might be better in this code path

Reply:
> From my understanding, a problem might occur if there is a `GetPods()` call when we have P3 in both maps. Which one should we return?
> 
> Considering it would happen only if, within 2 seconds, a pod is deleted and re-created with the same name and within the same namespace, I suggest returning nothing if the same pod is present in both maps. I prefer not to enrich things than enrich them wrongly.

Then we need to make sure that our gadgets are not prone to panic when it encounters a non enriched event. See top post of regarding `TestAdviseNetworkpolicy`

Reply:
I used `spec:  replicas: 300` in a test deployment:

> * if you add 3 containers to the cache in less than 0.5 seconds and the mutex causes a drop of performance, then we should avoid it.

No problems here

> * if you add 10000 containers in 0.5 seconds and no performance drop occurs, you can safely add a mutex.

It seems that we get `OnAdd` events in a ~20 pods per second rate. But this is also true even if the `OnAdd` does nothing (everything commented out)

> > How many container create and deletion events would happen concurrently? Meanwhile we have concurrent enrichments - of how many events per seconds?

There will never be any parallel Add/Update/Delete events. The interface says that this is the case and that the events are serialized for a single handler.
What we need to test is concurrent Add/Delete events while we enrich events. This can happen concurrently.